vm/proxyapp/proxyappclient.go

```diff
@@ -0,0 +1,469 @@
+
+	subStdin, err := subProcess.StdinPipe()
+	if err != nil {
+		freeAll()
+		return nil, fmt.Errorf("failed to get stdinpipe: %w", err)
+	}
+	toClose = append(toClose, subStdin)
+
+	subprocessLogs, err := subProcess.StderrPipe()
+	if err != nil {
+		freeAll()
+		return nil, fmt.Errorf("failed to get stderrpipe: %w", err)
+	}
+	toClose = append(toClose, subprocessLogs)
+
+	codec := jsonrpc.NewClientCodec(stdInOutCloser{
+		subStdout,
+		subStdin,
+	})
+
+	go io.Copy(os.Stdout, subprocessLogs)
+
+	err = subProcess.Start()
+	if err != nil {
+		freeAll()
+		return nil, fmt.Errorf("failed to start command %v: %w", cmd, err)
+	}
+
+	onTerminated := make(chan bool, 1)
+
+	go func() {
+		err := subProcess.Wait()
+		if err != nil {
+			log.Logf(0, "failed to Wait() subprocess: %v", err)
+		}
+		freeAll()
+		onTerminated <- true
+	}()
+
+	return &ProxyApp{
+		Client:       rpc.NewClientWithCodec(codec),
+		terminate:    cancelContext,
+		onTerminated: onTerminated,
+		subProcess:   subProcess,
+		params:       params,
+	}, nil
+}
+
+func (proxy *ProxyApp) CreatePool(config string, debug bool) (int, error) {
+	var reply proxyrpc.CreatePoolResult
+	err := proxy.Call(
+		"ProxyVM.CreatePool",
+		proxyrpc.CreatePoolParams{
+			Debug: debug,
+			Param: config,
+		},
+		&reply)
+	if err != nil {
+		return 0, err
+	}
+
+	return reply.Count, nil
+}
+
+func (proxy *ProxyApp) CreateInstance(workdir string, index int) (vmimpl.Instance, error) {
+	var reply proxyrpc.CreateInstanceResult
+	err := proxy.Call(
+		"ProxyVM.CreateInstance",
+		proxyrpc.CreateInstanceParams{
+			Workdir: workdir,
+			Index:   index},
+		&reply)
+	if err != nil {
+		return nil, fmt.Errorf("failed to proxy.Call(\"ProxyVM.CreateInstance\"): %w", err)
+	}
+
+	return &instance{
+		ProxyApp: proxy,
+		ID:       reply.ID,
+		params:   proxy.params,
+	}, nil
+}
+
+type instance struct {
+	*ProxyApp
+	ID     string
+	params *proxyAppParams
+}
+
+// Copy copies a hostSrc file into VM and returns file name in VM.
+// nolint: dupl
+func (inst *instance) Copy(hostSrc string) (string, error) {
+	var reply proxyrpc.CopyResult
+	err := inst.ProxyApp.Call(
+		"ProxyVM.Copy",
+		proxyrpc.CopyParams{
+			ID:      inst.ID,
+			HostSrc: hostSrc,
+		},
+		&reply)
+	if err != nil {
+		return "", err
+	}
+
+	return reply.VMFileName, nil
+}
+
+// Forward sets up forwarding from within VM to the given tcp
+// port on the host and returns the address to use in VM.
+// nolint: dupl
+func (inst *instance) Forward(port int) (string, error) {
+	var reply proxyrpc.ForwardResult
+	err := inst.ProxyApp.Call(
+		"ProxyVM.Forward",
+		proxyrpc.ForwardParams{
+			ID:   inst.ID,
+			Port: port,
+		},
+		&reply)
+	if err != nil {
+		return "", err
+	}
+	return reply.ManagerAddress, nil
+}
+
+func closeAll(closers []io.WriteCloser) {
+	for _, closer := range closers {
+		closer.Close()
+	}
+}
+
+func buildMerger(params *proxyAppParams, names ...string) (*vmimpl.OutputMerger, []io.WriteCloser, error) {
```

Comment:
This does not use proxyAppParams now and does not return an error.
Also pipes don't need to be closed, so it can return []io.Writer and we won't need freeAll calls on it.

At this point may be easier to open-code in instance.Run, but not sure.
```
	merger := vmimpl.NewOutputMerger(nil)
	receivedStdoutR, receivedStdoutChunks := io.Pipe()
	merger.Add("stdout", receivedStdoutR)
	receivedStderrR, receivedStderrChunks := io.Pipe()
	merger.Add("stderr", receivedStderrR)
	receivedConsoleR, receivedConsoleChunks := io.Pipe()
	merger.Add("console", receivedConsoleR)
```

Reply:
Nice cleanup.